### PR TITLE
Perf/prefix block key with blocknum

### PR DESCRIPTION
**Pull request by asdacap**
- Regardless of what kind of freezerdb/archive configuration, you'll need the block number.
- This add blocknumber to all method of BlockStore, and prefix the blockhash for blockdb with big endian block number.
- Interestingly, unlike receipts, it does not reduce write amplification much, probably due to inserting at both key range at the same time, probably due to much bigger value so should need much bigger file who knows, which is why I scrapped this before in favour of enabling blobfiles, which still performs orders of magnitude faster.
- It does however, have the advantage of being HDD friendly. And is should reduce compaction spike on full sync/normal operations.
- Graph is write amp on (blobfiles, no blocknumber prefix, with blocknumber prefix). Ignore write amp on blobfiles as the effective write bandwith for the compaction is arount 2MBs which is very low. The one with blocknumber prefix have high write amp on l1 for some reason.
Screenshot_2023-10-04_18-04-12

- Rpc performance does not seems to change much. Seems to have slightly higher throughput, but could be noise.
- Graph is (before, after, after without blobfiles)
Screenshot from 2023-10-05 19-13-42

- On HDD, with blobs turn on, it increase eth_getLogs (without filter) by about 2X. It seems that the blobs are still sorted but without compaction, it'll probably need to search multiple blob files. Disabling blob files improve it futher by another 2X. Does not seems to impact eth_getLogs with filter much. They are always slow.

- Graph is after, before.
Screenshot_2023-10-06_22-04-05

## Changes

- Prefix block key with block number.

#### What types of changes does your code introduce?

- [X] Optimization

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs [hunk 1]
```diff
@@ -1359,7 +1359,10 @@ private bool ShouldCache(long number)
                 return null;
             }
 
-            Block block = _blockStore.Get(blockHash, shouldCache: false);
+            BlockHeader header = FindHeader(blockHash, options);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/BlockTree.cs:1362]: Could we get just a number from the hash without getting whole header when combined with the second PR that prepends headers with block number and adds a second index?

File: src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs [hunk 2]
```diff
@@ -51,33 +52,73 @@ public void Insert(Block block)
...
 
-    public void Delete(Keccak blockHash)
+    private static void GetBlockNumPrefixedKey(long blockNumber, Keccak blockHash, Span<byte> output)
     {
-        _blockDb.Delete(blockHash);
+        blockNumber.WriteBigEndian(output);
+        blockHash!.Bytes.CopyTo(output[8..]);
+    }
+
+    public static byte[] GetBlockNumPrefixedKey(long blockNumber, Keccak blockHash)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs:68]: Does it has to be public?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/BlockTree.cs:1362]: Yea, thats another reason for the other PR.

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs [hunk 3]
```diff
@@ -293,6 +293,7 @@ public AddBlockResult Insert(BlockHeader header, BlockTreeInsertHeaderOptions he
                 throw new InvalidOperationException("Genesis block should not be inserted.");
             }
 
+            _headerStore.InsertBlockNumber(block.Hash, block.Number);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/BlockTree.cs:296]: reverse order? First set the data, then mapping -> cannot corrupt the mapping?

File: src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs [hunk 4]
```diff
@@ -51,33 +52,58 @@ public void Insert(Block block)
...
+        blockNumber.WriteBigEndian(output);
+        blockHash!.Bytes.CopyTo(output[8..]);
     }
 
-    public Block? Get(Keccak blockHash, bool shouldCache = false)
+    public void Delete(long blockNumber, Keccak blockHash)
     {
+        _blockDb.Remove(blockHash.Bytes);
+        _blockDb.Delete(blockNumber, blockHash);
+        _blockCache.Delete(blockHash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs:69]: reverse order of removal, first cache, then mapping, then data -> cannot corrupt anything even if something breaks.

Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs]: Done

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**